mscore/editstyle.cpp

```diff
@@ -676,9 +676,147 @@ EditStyle::EditStyle(Score* s, QWidget* parent)
+            { ElementType::TUPLET,              PageTuplets  },
+            { ElementType::ARPEGGIO,            PageArpeggios },
+            { ElementType::SLUR,                PageSlursTies },
+            { ElementType::SLUR_SEGMENT,        PageSlursTies },
+            { ElementType::TIE,                 PageSlursTies },
+            { ElementType::TIE_SEGMENT,         PageSlursTies },
+            { ElementType::HAIRPIN,             PageHairpins },
+            { ElementType::HAIRPIN_SEGMENT,     PageHairpins },
+            { ElementType::VOLTA,               PageVolta    },
+            { ElementType::VOLTA_SEGMENT,       PageVolta    },
+            { ElementType::OTTAVA,              PageOttava   },
+            { ElementType::OTTAVA_SEGMENT,      PageOttava   },
+            { ElementType::PEDAL,               PagePedal    },
+            { ElementType::PEDAL_SEGMENT,       PagePedal    },
+            { ElementType::TRILL,               PageTrill    },
+            { ElementType::TRILL_SEGMENT,       PageTrill    },
+            { ElementType::VIBRATO,             PageVibrato  },
+            { ElementType::VIBRATO_SEGMENT,     PageVibrato  },
+            { ElementType::BEND,                PageBend     },
+            { ElementType::TEXTLINE,            PageTextLine },
+            { ElementType::TEXTLINE_SEGMENT,    PageTextLine },
+            { ElementType::ARTICULATION,        PageArticulationsOrnaments },
+            { ElementType::FERMATA,             PageFermatas },
+            { ElementType::STAFF_TEXT,          PageStaffText },
+            { ElementType::TEMPO_TEXT,          PageTempoText },
+            { ElementType::LYRICS,              PageLyrics   },
+            { ElementType::LYRICSLINE,          PageLyrics   },
+            { ElementType::LYRICSLINE_SEGMENT,  PageLyrics   },
+            { ElementType::DYNAMIC,             PageDynamics },
+            { ElementType::REHEARSAL_MARK,      PageRehearsalMarks },
+            { ElementType::FIGURED_BASS,        PageFiguredBass },
+            { ElementType::HARMONY,             PageChordSymbols },
+            };
+
       cs->startCmd();
       }
 
+//---------------------------------------------------------
+//   gotoElement
+///   switch the page to the one related to the element `e`
+//---------------------------------------------------------
+
+void EditStyle::gotoElement(Element* e)
+      {
+      const ElementType& t = e->type();
+      const auto i = PAGES.find(t);
+      if (i != PAGES.cend())
+            setPage(pageStack->indexOf(i->second));
+      }
+
+//---------------------------------------------------------
+//    SUPPORTED_TYPES
+///    this is a static vector that must be kept in sync with the non-static
+///    member `pages`. The order is irrelevant, but the contents must be same as
+///    the keys of `pages`.
+//---------------------------------------------------------
+
+const std::vector<ElementType> EditStyle::SUPPORTED_TYPES = {
```

Comment:
The solution looks much better now but there is still a duplication of data between the `PAGES` map and `SUPPORTED_TYPES` array. The reason to have both is totally valid: it is indeed good to be able to know which elements have a relevant page without creating a whole `EditStyle` dialog instance. However, I believe if we go a bit further we can avoid a duplication even here: we can make `PAGES` a static member of `EditStyle` by using "pointer to member"s as its values. This would require us to explicitly dereference such values when using them but that would also allow us to define such values without having an actual instance of `EditStyle`. In the code that would be something like (I didn't try to compile this so some corrections may be needed):
```
// in EditStyle class declarations
typedef QWidget* EditStyle::* EditStylePage; // define an alias for "EditStyle member of QWidget* type"
static const std::map<ElementType, EditStylePage> pages;

// definition in .cpp
const std::map<ElementType, EditStylePage> EditStyle::pages = {
      { ElementType::SCORE,               &EditStyle::PageScore    },
      { ElementType::PAGE,                &EditStyle::PagePage     },
      // ...
      };

// gotoElement would contain something like
const auto i = pages.find(t);
if (i != pages.cend()) {
      QWidget* page = this->*(i->second);
      setPage(pageStack->indexOf(page));
      }
```

Then `elementHasPage` function could search the pages map directly thus eliminating a need for a separate `SUPPORTED_TYPES` list. This also doesn't lose any benefits of this solution that I mentioned earlier.

Reply:
Wow, that's interesting. I haven't ever used `typedef`s, so it's cool to see how that can be used. I can't see how it would work though... in the example, isn't `EditStyle::pages` a static map? If so, then how come it can access non-static members of `EditStyle` like `PageScore`? I don't really get it :/

Reply:
Yes, `pages` is a static map, but its values are assigned  not by non-static *members* of `EditStyle` but by [*pointers* to members](https://en.cppreference.com/w/cpp/language/pointer#Pointers_to_members) of `EditStyle` which are not bound to any particular instance of `EditStyle` class. Technically that could be implemented by compilers, for example, like storing a difference between a pointer to a member and a pointer to the whole `EditStyle` instance — that value does not (usually) depend on an instance being used.

In order to use this "pointer to member" value with a particular instance of `EditStyle` we have to use a [pointer-to-member access operator `->*`](https://en.cppreference.com/w/cpp/language/operator_member_access#Built-in_pointer-to-member_access_operators) which is done in this line of my example:
```
QWidget* page = this->*(i->second);
```
or, more explicitly,
```
EditStylePage pagePtr = i->second;
QWidget* page = this->*pagePtr;
```

In a similar way pointers to member functions can also be obtained which is used, for example, in the [new Qt 5's signal-slot connection syntax](https://wiki.qt.io/New_Signal_Slot_Syntax#New:_connecting_to_QObject_member).

Reply:
Ah, right. So it's pointers to the members of _any_ `EditStyle` instance, and when it's accessed as a member of an instance of `EditStyle` the pointers point to the actual objects. Nice! I'll try and implement that.

Reply:
Ok, how does it look now @dmitrio95 ?